Added logic to generate word suggestions and spellcheck

RadicalMuffinMan opened the pull request:
# Description

This PR implements a complete native Rust NLP engine for word suggestions and spell checking, it also addresses the memory constraints discussed in #677.

## Key Features
 - Native Rust Implementation
- All dictionary data and NLP processing runs in native memory, avoiding JVM heap limitations
- Trie-based data structure for efficient prefix matching and word lookup
- LZ4 compression for 40% smaller dictionary files with fast decompression (1GB/s)
- Language switching works seamlessly
- Personal dictionary persists across sessions and languages
- Multi-language word prediction based on the user's chosen subtypes, all languages are equal. Word prediction is based on the word's weighted frequency.
- Added n-grams support to help new users who haven't created a personal prediction library have easier use of the app
- Word predictions includes pronouns, acronyms, contractions, and french elisions 

## Dictionary Source
Word frequency lists sourced from [hermitdave/FrequencyWords](https://github.com/hermitdave/FrequencyWords) (MIT License), curated to remove offensive words while preserving dual-purpose terms (e.g., animal names that can be insults).

## Changes
### Kotlin

- LatinLanguageProvider.kt - Multi-language binary dictionary loading
- NlpBridge.kt - JNI bridge for native calls
- AppPrefs.kt - Added smartbar.autocorrect preference
- EditorInstance.kt - Auto-shift after sentence endings

### Rust

- nlp - Core NLP engine with trie, suggestions, spell check
- dict-compiler - Tool to compile JSON → binary dictionaries
- rust - JNI bindings

# APK testing

For each change in the pull request, a workflow is run, which produces a debug artifact APK. Go to Checks -> FlorisBoard CI -> `app-debug.apk` and download the APK. It installs under the `dev.patrickgold.florisboard.debug` namespace and will not mess with your main installation.

# Checklist

- [X] I have read and understood the [contribution guidelines](https://github.com/florisboard/florisboard/blob/main/CONTRIBUTING.md).
- [X] I have read and understood the [AI policy](https://github.com/florisboard/florisboard/blob/main/AI_POLICY.md).

Commit:
Added logic to generate word suggestions and spellcheck

patrickgold commented:
Hi, before I can take a closer look at your changes, what is the source for the data.json and under which license is it distributed originally?

Commit:
Enhance LatinLanguageProvider with context-aware suggestions and personal dictionary support

Commit:
Enhance suggestion handling and context awareness in LatinLanguageProvider

- Updated WordSuggestionCandidate to always be eligible for auto-commit.
- Improved commitChar logic to handle space input and auto-commit suggestions based on confidence.
- Introduced context-aware suggestion generation by extracting preceding words.
- Refactored spelling suggestion scoring to include frequency and context-based bonuses.
- Added phonetic indexing for fallback suggestions.
- Implemented methods for managing personal dictionary and context associations.
- Enhanced input shift state management for better user experience.

AxlNunez commented:
> Hi, before I can take a closer look at your changes, what is the source for the data.json and under which license is it distributed originally?

Originally I used the file that was already in the project, just reformatted it into proper json, capitalized some things, and it worked. Afterwards I felt like the list was too short so I sought out a bigger one and found it [here](https://github.com/dwyl/english-words?tab=Unlicense-1-ov-file#readme), which has an "unlimited license".

In the future I want to see if we can either separate each language into it's own separate json file or nest it under this main one.

patrickgold reviewed (commented):
Thanks for clarifying.

I have had a closer look at your proposed idea. It's major flaw is that it is implemented in Kotlin, which then in runtime runs in the JVM heap. While for normal logic and UI this is perfectly fine, it is a big issue for anything computationally and memory intensive. If you load the dictionary into a Kotlin map, it is stored in the JVM heap, which is quite limited. If you exceed the limit, it results in an `OutOfMemoryError`, which crashes the whole keyboard. Our first predictive text approach (a long time ago) was also written in Kotlin, so we have first-hand experience with these errors. See also the infamous https://github.com/florisboard/florisboard/issues/677 issue.

Our current approach is to implement it natively in Rust. Native has the big advantage that a) data can be stored with less overhead (no garbage collector metadata), and b) we just have a lot more space in native memory available. Also Unicode provides first-class ICU crates for text manipulation, which is a must for future-proofing the algorithm for non-English languages (think accents on letters, differing concepts in scripts, etc).

From your code I can see that you seem to have some experience with predictive text, or at least some idea on how to approach it. I also have some ideas how to approach it, but from a different angle, and with important constraints in mind. Maybe it would make sense if we get together and combine our ideas, what do you think?

RadicalMuffinMan committed:
Merge remote-tracking branch 'upstream/main' into roadmap/predictive-text

RadicalMuffinMan committed:
This replaces the previous Kotlin/JVM implementation with a native Rust predictive text engine, addressing feedback from PR #2970 and issue #677. The original approach risked OutOfMemoryError when loading large dictionaries; the new design stores all dictionary data in native memory, removing pressure from the Android JVM heap.

RadicalMuffinMan committed:
Enable suggestions in incognito mode and prevent learning during private sessions

RadicalMuffinMan committed:
Check autocorrect setting before returning auto-commit candidate

RadicalMuffinMan committed:
Updated auto-commit eligibility logic for suggestions based on confidence and frequency

RadicalMuffinMan committed:
- Trimmed the dictionary from 485K to 60K most common words
- Applied better rank-based frequency distribution (1-255 scale)
- Filtered offensive/inappropriate words
- Renamed data.json to en_US.json for multi-language prep
- Updated LatinLanguageProvider to load new dictionary

Overall the dictionary size was reduced from 9.4 MB to 1.02 MB (89% reduction)

RadicalMuffinMan committed:
Added binary trie implementation for efficient word storage and retrieval

RadicalMuffinMan committed:
Added dict-compiler for compiling JSON dictionaries to binary trie format

RadicalMuffinMan committed:
Implemented binary dictionary loading and fixed the NLP engine with typo correction features

RadicalMuffinMan committed:
Refactored NLP engine to support language-specific dictionary loading and management

RadicalMuffinMan committed:
Added LZ4 compression to binary dictionary format

- Compressed trie node data with LZ4 for ~43% smaller dictionary files
- Updated binary format version to 2
- Sorted words by frequency before trie insertion for better locality

RadicalMuffinMan committed:
Added support for German, Spanish, French, Italian, and Portuguese

RadicalMuffinMan commented:
@patrickgold it's ready, friend.

patrickgold commented:
Thanks a lot for the changes!

I am really impressed, my first impression is quite good! We are currently doing an initial internal review of your changes and will come back to you.

Some questions for now:
1. The NLP engine seems to support current word prediction and spell check. Would the implementation be extensible in the future to also support next word prediction? Or would we need a separate NLP engine for that?
2. How difficult would it be to extend the NLP engine to support multi-language prediction in the future? E.g. en+de at once.
3. The personal dictionary/learning is shared between languages, is this intended?
4. You seem to be quite talented in this area of development. Would you be interested to stay around to develop/maintain NLP-related features in FlorisBoard?

RadicalMuffinMan committed:
Implemented next word prediction feature in NLP engine

RadicalMuffinMan committed:
Add n-grams support for predictive text in NLP engine and added Latin America spanish language support

RadicalMuffinMan commented:
@patrickgold so I added predictive text support but wanted to go one step further:

How Next-Word Prediction Worked

**Without N-grams:**
- User types "how " --> Keyboard has NO predictions until user teaches it
- User has to type "how are" multiple times before keyboard learns the pattern
- Cold start problem: new users get zero predictions

**With N-grams:**
- User types "how " --> Keyboard immediately suggests "are", "is", "do" from n-gram data
- Works from first use, no learning required
- As user types more and more, learned patterns override/supplement n-grams

## Data Flow

```
User types "how " (space pressed)
            ↓
    predict_next_word(["how"])
            ↓

┌────────────────────────────────────┐
│  1. Check learned associations     │  ← User's personal patterns (10x weight)
│     context_map["how"] → {"are": 5}│
│                                    │
│  2. Check n-gram data              │  ← Pre-built baseline (1x weight)
│     ngrams["how"] → {"are": 100,   │
│                      "is": 90, ...}│
└────────────────────────────────────┘

            ↓
Merge & rank by weighted score
            ↓
Return: ["are", "is", "do", ...]
```
    
## Priority System

| Source | Weight | Purpose |
|--------|--------|---------|
| Learned (context_map) | 10x | User's personal typing patterns |
| N-grams | 1x | Baseline predictions for everyone |

So after typing "how are" 2-3 times, your learned "are" will outrank any n-gram suggestion.

## Files Involved

- **language.ngrams.json** --> Loaded at startup, provides instant predictions
- **context_map.json** --> Persisted to disk, stores your learned patterns
- Both merge in `predict_next_word()` with learned data prioritized

The n-grams are basically like  a "starter pack" so the keyboard works intelligently from day one, then adapts to each user's style over time.

Now I'm trying to fix contractions and pronouns because it's annoying me. Also I added a Latin America Spanish dictionary

RadicalMuffinMan committed:
Added canonical form support for contractions, acronyms, and proper nouns

Typing "im" now suggests "I'm", "usa" suggests "USA", and "germany"  suggests "Germany" with auto-commit enabled.

- Extended binary dictionary format to v3 with embedded canonical forms
- Added direct canonical lookup in suggest() for exact matches
- Extracted canonical forms during dictionary compilation
- Updated spell checker to recognize canonical mappings

RadicalMuffinMan commented:
All right, pronouns, acronyms, contractions, and french elisions are added

RadicalMuffinMan committed:
- Refactored NLP engine to support multiple active languages and updated related methods for language management
- Language used is based on subtypes added

patrickgold commented:
Many thanks for your recent changes, I am really impressed! Will have a look at the changes later today.

> Did you want me to try and add a json and dict for all languages available in the subtypes?

I would argue no, except if you have a good data source to create a new dictionary for an additional language. Other languages can always be added as new dictionary extensions.

Some of the languages (e.g. Chinese, ...) require highly specialized engines anyways, thus dictionaries for these languages would make no sense.

RadicalMuffinMan committed:
Removed all non US English dictionary files from assets

RadicalMuffinMan commented:
I also made another PR that fixes the app crashing on banking apps due to the numeric input

patrickgold commented:
Quick question regarding the dictionary/json files: as far as I see, the binary dict files have a magic header for identification and version field for forward compatibility, while the ngrams.json doesn't. Is this intentional? I am asking because we are currently drafting the dictionary decl format within extension formats and need to decide if the extension.json needs to handle schema versions or if the dictionary files themselves handle this.

patrickgold commented:
Thanks for the quick reply. Our current draft for the new `keyboard3` extension (which unifies the existing `keyboard` and `languagepack` extensions) would define Latin dictionaries like this:


```jsonc
{
  "$": "ime.extension.keyboard3", // new ID, will replace current keyboard and languagepack formats in the future
  "meta": {
    "id": "org.florisboard.languages.en_US",
    "version": "0.1.0",
    "title": "Language Pack English (US)",
    "description": "yadayada",
    "maintainers": [ "author1 <author1@example.com>" ],
    "license": "MIT"
  },
  "dictionaries": [
    {
      "$": "ime.dictionary.latin",
      "lang": "en_US",
      "dictVersion": 1,
      "dictPath": "dictionaries/en_US.dict",
      "ngramsVersion": 1,
      "ngramsPath": "dictionaries/en_US.ngrams.json"
    },
  ]
}
```

The `"$": "ime.dictionary.latin"` ID allows us to distinguish between dicts targeting the Latin NLP engine and dicts for other NLP engines.

With this approach the versions are declared in the extension manifest, and will then be passed along to the NLP engine.

RadicalMuffinMan committed:
Added support for dynamic text casing in word suggestions based on input shift state

RadicalMuffinMan committed:
Updated the NLP engine to preserve case for acronyms and add support for cases like lowercase "us" and "add" not changing to "US" (country) and "ADD" (medical condition) in suggestions

patrickgold commented:
Thanks for sharing the resources used for the implementation, and admitting the usage of generative AI (Claude) for parts of the code.

Please note that the use of generative AI is not something we see positively, as stated in the [AI policy](https://github.com/florisboard/florisboard/blob/main/AI_POLICY.md) you agreed to in the original PR comment. It adds additional overhead to us, **human** reviewers, to not only check the code for potential issues and functionality, but also authenticity, potential copyright infringement, and missing attribution caused by AI. We do not know how similar your approach is to e.g. LatinIME or the Chromium spell checker dictionaries you listed, and if that necessitates attribution. You were able to list the sources used, which is a big plus, and we think in general you have the necessary skills and technical knowledge to develop code yourself, however parts of your submitted code and some of the comments in this reply chain are heavily influenced by generative AI, which is a violation of our AI policy.

We have discussed internally how to proceed, and think that the general idea and approach of the word prediction engine is okay and in simple tests it performs well, however given that AI was involved and left its traces the current state of the PR is not merge-able. We thus propose the following:

- You go over the code and ensure that, where similar approaches are taken, proper attribution to the respective open-source project/paper is given
- You ensure that the code is cleaned from "AI artifacts" and fully generated AI files are removed / replaced by your own code
- Further contributed code is not the product of generative AI.

I hope you can understand our standpoint and why we need to enforce the policy.
